### ISSUE TITLE
valid-jsdoc fails to validate return of callbacks (with requireReturn: false)

### ISSUE THREAD

**Issue by fcanela** [eslint/eslint#5207]
**Eslint version**: v1.10.3

# What I did

* I enabled valid-jsdoc and set "requireReturn" as false.
````
valid-jsdoc: [1, { "requireReturn": false }]
````
* I linted the following code
````js
/**
 * Close the web server
 *
 * @param {function} callback - Called after web server is stopped
 */
PolyApp.prototype.stop = function(callback) {
  if (!this._listeningServer) {
    if (callback) {
      return callback();
    }

    return;
  }

  this._listeningServer.close(callback);
};
````
# What happened

It failed to validate asking me to add a return field in JsDoc.
> warning  Missing JSDoc @returns for function  valid-jsdoc

# Expected result

Code being valid as this return was a for finishing the execution instead of giving back any result (validating callback-return rule). No return value should be expected from this method.

The second return produces no warning. I think the first one is having the same behaviour as the second, so it should produce no warning too.

# Complete output
````
> eslint -c .eslintrc lib/*


/home/fcanela/workspace/funbox/polyapp/lib/polyapp.js
  50:1  warning  Missing JSDoc @returns for function  valid-jsdoc

✖ 1 problem (0 errors, 1 warning)
````

**Comment by eslintbot** [comment]
@fcanela Thanks for the issue! If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

**Comment by MathRobin** [comment]
As you are returning something (you're using return), it's normal that you have to provide return comment, no?

**Comment by fcanela** [comment]
Hello, Mathieu. Thanks for engaging.

From my point of view this return behaves like a break more than passing information to the caller. Sometimes functions that returns a callback can also return nothing (undefined), like in my example.  

Should I then mark that return as optional to document the real behaviour? Am I able to do that with a "return" statement? I was unable to find how in a quick Google search.

Given a function accepts a callback and returns nothing (undefined) except, sometimes, the callback. Is there a context in which not documenting this function could lead to any problem? Is there any context in which a function that sometimes returns a callback value could break something? Do anyone have a common case of returning useful values in callbacks which should be returned as well?

**Comment by ilyavolodin** [comment]
Sorry, I'm not a jsDoc expert, but I think the way you do what you described is:
```js
@returns {function|undefined}
```
Although I might be wrong

**Comment by fcanela** [comment]
You are right, @ilyavolodin: multiple return types could be expressed that way.

I opened a [Stack Overflow discussion](http://stackoverflow.com/questions/35343952/should-i-document-a-return-callback-as-return-value-if-it-is-not-intended-to) to gather some feedback.

If there is a good reason to force developers to document "return" even when working as flow control mechanism we can close this issue. If not, I am open to help developing the solution as it result in unnecessary warnings in my projects.

**Comment by ilyavolodin** [comment]
@fcanela You can also always specify `requireReturn: false` [option](http://eslint.org/docs/rules/valid-jsdoc.html#requirereturn) and the rule will stop flagging missing returns. I think that's what you probably want to do, if specifying multiple return types isn't an option.

**Comment by fcanela** [comment]
As I wrote in the report the problem is that the warning still appears after disabling `requireReturn`. I may have not configured it propertly, that's why I included the configuration string as said in the Contributors Code.

**Comment by ilyavolodin** [comment]
Ahh.. Sorry, I didn't notice that. OK, so that might be a bug in that case. @eslint/eslint-team I'm not very familiar with this rule, but based on the documentation if `requireReturn: false` is specified, you shouldn't see this error, correct?

**Comment by fcanela** [comment]
If `requireReturn: false` is specified you only have to document "return" if there is any "return" in the code.

If you place "return;" you have to document nothing, but if you place "return callback();" then you have.

I can think some solutions: 

* Enabling a "never" option so you are never asked for "return".
* Including "return callback();" in the list of exceptions (could require more changes).
* Creating another option to disable requirements for callbacks.

**Comment by mysticatea** [comment]
As far as I know, `requireReturn` option behaves as like below:

**In true case:**

```js
/**
 * test.
 * @returns {void}
 */
function foo() {
}

/**
 * test.
 * @returns {number} 1.
 */
function bar() {
    return 1;
}
```

**In false case:**

```js
/**
 * test.
 */
function foo() {
}

/**
 * test.
 * @returns {number} 1.
 */
function bar() {
    return 1;
}
```

And that function is returning a value (the result of `callback()`), so it's intentional behavior.

----

I don't think making an exception here is good because this is too special. There are callbacks which returns a meaningful value such as `Array.prototype.filter`, `map`, and `reduce`.
Also, I believe the option of not writing `@return` is bad idea because it's opposite from the purpose of this rule.

If ESLint gets Type Inference feature, perhaps we can handle this correctly.
But unfortunately, ESLint doesn't have the feature currently.

In my opinion, you can modify the code simply as below for now.

```js
callback();
return;
```

**Comment by fcanela** [comment]
Thanks for your point of view, @mysticatea. After thinking a little bit those are my thoughs:

Those examples you proposed as "callbacks" aren't callback in the meaning of "ping me when the job is done", are utility functions which perform a work needed for the calling function. As it, they should not be called "callback" or "cb" as parameters and we can differentiate them doing the same inference that "callback-return" rule does.

I would love to modify the code to be as you suggested but that would conflict with *callback-return* and it benefits.

I would love to be able to benefit from both rules without having to document return values which I do not want to compromise with and will fill both code and documentation without providing value to the user.

I do not intend to be pushy with this issue. I am only trying to express my thoughs because I think the have a reason to be. If you decide that the project should not implement this (even though I proposed myself to help and try a PR) to avoid features cluttering I will understand.

**Comment by mysticatea** [comment]
Thanks @fcanela.

Indeed, my examples were not "callback".
Hmm, about an option for callbacks, I will follow owner/reviewer's direction.

Just FYI: Actually `callback-return` rule allows the following pattern, I conformed it on the online demo:

```js
/*eslint callback-return: 2*/

function foo(callback) {
    if (a) {
        callback();
        return;
    }
}
```

http://eslint.org/demo/

**Comment by fcanela** [comment]
Ah, I missed that somehow. Thanks for pointing, @mysticatea.

I like to place the returns with the callback to avoid human errors but having eslint reduces the chances of that happenings and your solution clearly solves the problem with breaking both rules and having extra warnings.

I will wait too for a eslint team decision. If they think this is too granular to be implemented I will follow the style you implemented in the demo. Thanks again!

**Comment by fcanela** [comment]
I worked a little bit with that style and I think that it leads to a lot of code for a simple issue.

For this:
```js
if (!this._listeningServer) {
  if (callback) return callback();
  return;
}
```
I would have to code this so no eslint issue is arose:
````js
  if (!this._listeningServer) {
    if (callback) { 
      callback();
      return;
    }
    return;
  }
````

If the team accepts this behavior as  acceptable I will disable the callback return rule and go just with:
```js
if (!this._listeningServer) {
  if (callback) callback();
  return;
}
```
But I would prefer to benefit from this eslint rule without having to do weird transformations to my code.

**Comment by nzakas** [comment]
I'm sorry, this is too specific an exception to add. If you're looking for an alternative, try https://github.com/gajus/eslint-plugin-jsdoc.
